File: day_4/solve.py
import copy
import sys
from os import read
from itertools import chain


def put_str_to_nums(str):
    return [x for x in str.split(" ") if x != r""]


def compose_boards(lines):
    boards = []
    step = -1

    for line in lines:
        if line == "":
            boards.append([])
            step += 1
            continue
        boards[step].append(put_str_to_nums(line))

    return boards


def fill_in_board(board, pick):
    for r in range(len(board)):
        for c in range(len(board[r])):
            if board[r][c] == pick:
                return [True, [r, c]]
    return [False]


def rotate_board_90(A):
    N = len(A[0])
    for i in range(N // 2):
        for j in range(i, N - i - 1):
            temp = A[i][j]
            A[i][j] = A[N - 1 - j][i]
            A[N - 1 - j][i] = A[N - 1 - i][N - 1 - j]
            A[N - 1 - i][N - 1 - j] = A[j][N - 1 - i]
            A[j][N - 1 - i] = temp

    return A


def check_for_winner(board):
    for row in board:
        if all(i == "X" for i in row):
            return [True, board]

    rotated_board = rotate_board_90(copy.deepcopy(board))
    for row in rotated_board:
        if all(i == "X" for i in row):
            return [True, board]

    return [False]


def run(FILE_NAME):
    f = open(FILE_NAME, "r")
    text = f.read()
    lines = text.split("\n")
    draws = lines[0].split(",")
    mut_boards = compose_boards(lines[1:])

    for draw in draws:
        for i in range(len(mut_boards)):
            mut_board = mut_boards[i]
            valid_draw = fill_in_board(mut_board, draw)

            if not valid_draw[0]:
                continue

            [row, col] = valid_draw[1]
            mut_board[row][col] = "X"

            result = check_for_winner(mut_board)

            if result[0]:
                winner = result[1]
                N = len(winner)
                tot = sum(
                    [
                        int(winner[x][y])
                        for x in range(N)
                        for y in range(len(winner[x]))
                        if winner[x][y] != "X"
                    ]
                )
                print(tot * int(draw))
                return


if __name__ == "__main__":
    if len(sys.argv) > 1:
        run("sample.txt")
    else:
        run("input.txt")

File: day_4/solve.js
function formatInput(rawInput) {
  const input = rawInput.split("\n\n").filter(Boolean);
  const numbers = input[0].split(",").map((n) => +n);
  const boards = input.slice(1).map((grid) =>
    grid
      .split("\n")
      .filter(Boolean)
      .map((v) =>
        v
          .split(" ")
          .filter(Boolean)
          .map((n) => ({ value: +n, marked: false }))
      )
  );

  return { numbers, boards };
}
function solution1({ numbers, boards }) {
  const sum = (board) => {
    let s = 0;
    for (const c of board) for (const x of c) s += !x.marked ? x.value : 0;
    return s;
  };

  for (const n of numbers) {
    for (const board of boards) {
      for (const row of board) {
        for (const col of row) {
          if (col.value === n) {
            col.marked = true;
          }
        }
      }
      let won = false;
      for (let i = 0; i < board.length; i++) {
        let found = true;
        for (let j = 0; j < board[i].length; j++) {
          if (!board[i][j].marked) {
            found = false;
            break;
          }
        }
        if (found) {
          won = true;
          break;
        }
      }
      for (let i = 0; i < board.length; i++) {
        let found = true;
        for (let j = 0; j < board[i].length; j++) {
          if (!board[j][i].marked) {
            found = false;
            break;
          }
        }
        if (found) {
          won = true;
          break;
        }
      }
      if (won) return sum(board) * n;
    }
  }
}
function solution2({ numbers, boards }) {
  const sum = (board) => {
    let s = 0;
    for (const c of board) for (const x of c) s += !x.marked ? x.value : 0;
    return s;
  };

  const winning = [];
  for (const n of numbers) {
    for (const [k, board] of boards.entries()) {
      if (winning.includes(k)) continue;
      for (const row of board) {
        for (const col of row) {
          if (col.value === n && !col.marked) {
            col.marked = true;
          }
        }
      }
      let won = false;
      for (let i = 0; i < board.length; i++) {
        let found = true;
        for (let j = 0; j < board[i].length; j++) {
          if (!board[i][j].marked) {
            found = false;
            break;
          }
        }
        if (found) won = true;
      }
      for (let i = 0; i < board.length; i++) {
        let found = true;
        for (let j = 0; j < board[i].length; j++) {
          if (!board[j][i].marked) {
            found = false;
            break;
          }
        }
        if (found) won = true;
      }
      if (won) {
        if (boards.length - winning.length > 1) winning.push(k);
        else return sum(board) * n;
      }
    }
  }
}

console.log(solution1(formatInput(input)));
console.log(solution2(formatInput(input)));
